sql: add error handling to the connExecutor constructor

knz opened the pull request:
This plumbs errors during construction of the connExecutor into a pgwire error visible to the client.

Right now errors can never happen so there is no way to observe the result of thsi change. It is a prefix for a larger PR which will exercise this code path.

knz committed:
sql: add error handling to the connExecutor constructor

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30691)
<!-- Reviewable:end -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/conn_executor.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQibVJ3kS0P1j431gB:-LNQibVJ3kS0P1j431gC:b-coydl) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L399)):*
> ```Go
> // memMetrics: The metrics that statements executed on this connection will
> //   contribute to.
> func (s *Server) SetupConn(
> ```

I don't quite understand the need for this dance. Why can't `ServeConn` just write whatever error it wants directly to the connection (instead of lifting that responsibility to `conn`)? Since obviously it is generally the sql layer that writes errors on a connection. Export a version of `conn.writeErr` if you need. Wouldn't that be better?

And btw, when introducing new places that can return errors to the client, be careful if the protocol supports an error at that particular place. We already apparently have a bug where we return an error when the protocol doesn't allow (or at least when lib/pq doesn't like it) - #24149. Unrelated, if you happen to want to dig into that  bug too... :P
___
*[pkg/sql/pgwire/conn.go, line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQcDtnBGXH7jYkiwN1:-LNQcDtnBGXH7jYkiwN2:bxxpnn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/075b8645627afa6e13e62a1db6a17789c2731786/pkg/sql/pgwire/conn.go#L238)):*
> ```Go
> 	}
> 
> 	ctx = logtags.AddTag(ctx, "user", c.sessionArgs.User)
> ```

what's up with this change?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/conn_executor.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQibVJ3kS0P1j431gB:-LNQsgMP6PxrnaLevvuE:b3gkm3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L399)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't quite understand the need for this dance. Why can't `ServeConn` just write whatever error it wants directly to the connection (instead of lifting that responsibility to `conn`)? Since obviously it is generally the sql layer that writes errors on a connection. Export a version of `conn.writeErr` if you need. Wouldn't that be better?

And btw, when introducing new places that can return errors to the client, be careful if the protocol supports an error at that particular place. We already apparently have a bug where we return an error when the protocol doesn't allow (or at least when lib/pq doesn't like it) - #24149. Unrelated, if you happen to want to dig into that  bug too... :P
</blockquote></details>

If you peek at #30702 you will understand this change is meant to enable the SessionData initialized by SetupConn in the output of server parameters just below.

Does this answer your question?
___
*[pkg/sql/pgwire/conn.go, line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQcDtnBGXH7jYkiwN1:-LNQsoN4AThFJ3fuSIow:b-5mzdxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/075b8645627afa6e13e62a1db6a17789c2731786/pkg/sql/pgwire/conn.go#L238)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's up with this change?
</blockquote></details>

Ensure that the events generated by the call to SetupConn above have the username logged as well. Is that what you are asking?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/conn_executor.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQibVJ3kS0P1j431gB:-LNRZwAz0qLwJFq8dlR3:bx8i10e) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L399)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If you peek at #30702 you will understand this change is meant to enable the SessionData initialized by SetupConn in the output of server parameters just below.

Does this answer your question?
</blockquote></details>

it only makes me wonder more why this commit has been isolated :P

What I think I would do is extract the `SessionData` construction out of the `connExecutor` ctor and do it before the `connExecutor` is created. Have you tried that?


<!-- Sent from Reviewable.io -->

knz commented:
> What I think I would do is extract the SessionData construction out of the connExecutor ctor and do it before the connExecutor is created. Have you tried that?

I'll try that. Thanks for the suggestion.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/conn_executor.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQibVJ3kS0P1j431gB:-LNVtbfm2flZw6zd3RXR:buiz87i) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L399)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it only makes me wonder more why this commit has been isolated :P

What I think I would do is extract the `SessionData` construction out of the `connExecutor` ctor and do it before the `connExecutor` is created. Have you tried that?
</blockquote></details>

Andrei I have added a 2nd commit which does what you suggest. You can compare the two approaches by either looking at the diff for the 1st commit, or the diff for the full PR (the diff from 1st to 2nd commit is perhaps not so interesting).

Which approach do you prefer? I think the first one was a bit cleaner, because in the 2nd approach which you suggest the `sessionDataMutator` is not in a "clean" state between the two phases. But I don't really care either way, please let me know how to proceed.

Cheers


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
as discussed, let's go with the original approach as I've got another use case in mind that would benefit from the connExecutor being created early and so I've conveniently changed my stance

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LNplCsO62j_sSRhESsw:b-a4c3sg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpk7xP-X5v3dHc0dUd:-LNpk7xP-X5v3dHc0dUe:bt4g3me) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L389)):*
> ```Go
> }
> 
> // SetupConn creates a connExecutor for the client connection.
> ```

pls give a hint that the result is supposed to be passed to `ServeConn` and also give a nod that this validates the `SessionArgs`
___
*[pkg/sql/conn_executor.go, line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpjoSW21xUx4wdttvj:-LNpjoSW21xUx4wdttvk:bobttfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L415)):*
> ```Go
> // and the ServeConn below. It encapsulates the connExecutor and hides
> // it away from other packages.
> type ConnectionHandler *connExecutor
> ```

I was convinced in another instance when I was trying to use a type alias similarly a struct that contains the `*connExecutor` as an unexported field is a better idea because it doesn't expose the exported `connExecutor` methods to the users in other packages.
___
*[pkg/sql/conn_executor.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpl-eY61FZJZBWJyEU:-LNpl-eY61FZJZBWJyEV:b-w578ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L419)):*
> ```Go
> // ServeConn serves a client connection by reading commands from
> // the stmtBuf embedded in the connHandler.
> func (s *Server) ServeConn(
> ```

I'd make this a method on the `ConnectionHandler`


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Made the requested changes, ready for another (hopefully last) look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LNpveSR0ikUyEtxNPS0:bnpca58)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpk7xP-X5v3dHc0dUd:-LNpvW9u6YcG7KKZUkhe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L389)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls give a hint that the result is supposed to be passed to `ServeConn` and also give a nod that this validates the `SessionArgs`
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNQibVJ3kS0P1j431gB:-LNpv4hV9-owbzwAuzo1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L399)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Andrei I have added a 2nd commit which does what you suggest. You can compare the two approaches by either looking at the diff for the 1st commit, or the diff for the full PR (the diff from 1st to 2nd commit is perhaps not so interesting).

Which approach do you prefer? I think the first one was a bit cleaner, because in the 2nd approach which you suggest the `sessionDataMutator` is not in a "clean" state between the two phases. But I don't really care either way, please let me know how to proceed.

Cheers
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpjoSW21xUx4wdttvj:-LNpv3ZP4uR2dwwFW9kO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L415)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I was convinced in another instance when I was trying to use a type alias similarly a struct that contains the `*connExecutor` as an unexported field is a better idea because it doesn't expose the exported `connExecutor` methods to the users in other packages.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpl-eY61FZJZBWJyEU:-LNpv4AH1zj_Qdir95N0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L419)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd make this a method on the `ConnectionHandler`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpk7xP-X5v3dHc0dUd:-LNpz1De1mBW9cr39MdD:b-7ejnf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L389)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

is this comment about the resources being released true? What resources - does this open a monitor? Can that be moved to `ServeConn()` - cause I think it'd be nice if we didn't mandate that `ServeConn` is called...
Related - it's unclear what the contract is when an error is returned. This comment and the fact that we're returning a non-empty value on error suggests that `ServeConn()` still needs to be called. I think we should return an empty value on error (and also make sure that no resources leak in that case).


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpk7xP-X5v3dHc0dUd:-LNqm24s6pRz42zn64H4:b-nccczj) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L389)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is this comment about the resources being released true? What resources - does this open a monitor? Can that be moved to `ServeConn()` - cause I think it'd be nice if we didn't mandate that `ServeConn` is called...
Related - it's unclear what the contract is when an error is returned. This comment and the fact that we're returning a non-empty value on error suggests that `ServeConn()` still needs to be called. I think we should return an empty value on error (and also make sure that no resources leak in that case).
</blockquote></details>

Andrei your comment is awesome. This is exactly the kind of scrutiny I was looking for in a review!

I have detailed the comment to explain why a close is necessary (and that this close is performed by ServeConn) -- the eventLog needs Finish() and some memory monitors + bound account need closing.


<!-- Sent from Reviewable.io -->

knz commented:
@andreimatei let me know when you're good with this so I can go forward and merge #30702 instead (so that both this change and the code that actually does something useful with it get shipped together)

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNpk7xP-X5v3dHc0dUd:-LNuQmDS7luIE5HWnl6-:br7ap5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/be6821dc0196bd12db797f84331e7d5d6cad4a17/pkg/sql/conn_executor.go#L389)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Andrei your comment is awesome. This is exactly the kind of scrutiny I was looking for in a review!

I have detailed the comment to explain why a close is necessary (and that this close is performed by ServeConn) -- the eventLog needs Finish() and some memory monitors + bound account need closing.
</blockquote></details>

Looking at #30702, it seems to me that `newConnExecutor()` starts monitors before returning an error - even though the contract of `SetupConn()` now states that the caller doesn't have to do anything if an err is returned. Isn't that a leak?

Also, what's `CloseBeforeUse()`? I don't see it here or in #30702. And do we really need it?


<!-- Sent from Reviewable.io -->

knz commented:
So yeah no matter how hard I was pulling at this it's impossible to have a clean resource policy as long as SetupConn() is activating monitors / event logging etc.

So I solved as follows:

- changed the contract of SetupConn to never engage resources no matter what
- added a separate method in between newConnEx() and run() called activate(), which starts the monitors and create the event trace.
- documented the things.

Can you check this again?

andreimatei reviewed (changes requested):
Why do we need the `activate()` - why can't `run ()` do whatever `activate()` is doing?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LNzjChg5z83lhhvQmUS:b-v8vqqi)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 537 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LNziz26DPWcmWMZ0KhY:-LNziz26DPWcmWMZ0KhZ:baob4cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0062df8986fcef31e25ee8de228b52ba9d140359/pkg/sql/conn_executor.go#L537)):*
> ```Go
> 
> 		// ctxHolder will be reset at the start of run(). We only define
> 		// it here so that CloseNoUse() doesn't panic.
> ```

Is `CloseNoUse` still a thing?


<!-- Sent from Reviewable.io -->

knz commented:
run() can't do the thing because newConnExecutorInTxn(). This needs to call resetInTxn() which needs an activated executor.

Can you check?
-- 
Raphael 'kena' Poss
Sent from mobile. Excuse my conciseness.

knz commented:
Also please step in and show me what you want me to write. You're clearly more experienced than me and I have a hard time understanding what you want me to do. 

Just asking "why why why" is not actionable: obviously I'm doing this change because it enables me to do something previously impossible. But there are 1000 ways to achieve that and you won't get me anywhere asking why I choose 1 over the 999 others. I just happens to work!!!


-- 
Raphael 'kena' Poss
Sent from mobile. Excuse my conciseness.

andreimatei reviewed (commented):
I don't know exactly what to write... 
I think it's pretty clear that a contract that says "after you call this, you have to call `activate` and then `run`" is dubious - I think the benefits of simple contracts are obvious and so the question about why can't `run()` just do the "activation" part seems legitimate to me. https://www.youtube.com/watch?v=uAb3TcSWu7Q
Lacking good reason, a client shouldn't have to understand anything about this activation.

Now,  I think `newConnExecutorWithTxn()` is a weirdo needed by the `InternalExecutor`; we shouldn't let this tail wag the dog; the interactions between the `conn` and the `connExecutor` are much more important in my opinion. So how about making `run()` check if it was already `activate()d` (probably by checking a new bool flag), and if it hasn't, do it. Then, `newConnExecutor()`'s contract can be "after calling this, you can inspect session variables and do whatever conn needs to do with them, and then you *can* call run()".
`newConnExecutorWithTxn()` can call `activate()`; that guy  already has a more involved contract - `close()` needs to be called on that `connExecutor` no matter what. But that's done, and limited to, the `InternalExecutor` which is a bit convoluted by its nature (or perhaps by the limits of my abilities :P).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LO-1Niw3R0AhS94-ZOt:b-eef41q)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

knz commented:
Ok so you're saying "keep activate() but have run() do it if newConnExecutorIntxn() hasn't called it already" -- am I understanding this right?

andreimatei reviewed (commented):
yes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LO-HvIz6HnFKRDSkqyS:b2nign6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

knz commented:
Done, PTAL

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-:-LO-dtCB8cq_MvubE-EK:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 1032 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LO-dY0L0AD13go9Tejv:-LO-dY0L0AD13go9Tejw:b-qheib7) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f12428fc0e5338c426be24986e4221fe887b99/pkg/sql/conn_executor.go#L1032)):*
> ```Go
> // activate engages the use of resources that must be cleaned up
> // afterwards. after activate() completes, the close() method must be
> // called even in error cases.
> ```

nit: "even in error cases" doesn't seem to make sense, or rather it's unclear whose errors this is talking about. I think saying that `close()` needs to be called is sufficient,


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30691)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/conn_executor.go, line 1032 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30691#-LO-dY0L0AD13go9Tejv:-LO2MomeD_WBoDdpbmTl:b24vfde) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f12428fc0e5338c426be24986e4221fe887b99/pkg/sql/conn_executor.go#L1032)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: "even in error cases" doesn't seem to make sense, or rather it's unclear whose errors this is talking about. I think saying that `close()` needs to be called is sufficient,
</blockquote></details>

Good suggestion! Thank you. Done.


<!-- Sent from Reviewable.io -->

knz commented:
Merged in  #30702
